core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1760,25 +1760,24 @@ class SparkContext(config: SparkConf) extends Logging {
   def listJars(): Seq[String] = addedJars.keySet.toSeq
 
   /**
-   * Shut down the SparkContext.
+   * When stopping SparkContext inside Spark components, it's easy to cause dead-lock since Spark
+   * may wait for some internal threads to finish. It's better to use this method to stop
+   * SparkContext instead.
    */
-  def stop(): Unit = {
-    if (env.rpcEnv.isInRPCThread) {
-      // `stop` will block until all RPC threads exit, so we cannot call stop inside a RPC thread.
-      // We should launch a new thread to call `stop` to avoid dead-lock.
-      new Thread("stop-spark-context") {
-        setDaemon(true)
+  private[spark] def stopInNewThread(): Unit = {
+    new Thread("stop-spark-context") {
+      setDaemon(true)
 
-        override def run(): Unit = {
-          _stop()
-        }
-      }.start()
-    } else {
-      _stop()
-    }
+      override def run(): Unit = {
+        SparkContext.this.stop()
```

Comment:
> Will this ever throw an exception? Should we register an UncaughtExceptionHandler or try catch with logging?

This happens in the driver, so we cannot use `SparkUncaughtExceptionHandler` to catch the error. The error will be sent to the user's UncaughtExceptionHandler if specified or just print to stderr.